server/scrapers.ts

```diff
@@ -0,0 +1,216 @@
+        const semester = semesterData.semester;
+        const subjects = semesterData.subjects;
+
+        return Promise.all(subjects.map(async subject => {
+            const classesQuery = await axios.get("https://classes.cornell.edu/api/2.0/search/classes.json?roster=" + semester + "&subject=" + subject.value, { timeout: 30000 });
+            if (classesQuery.status !== 200) {
+                console.log("Unable to fetch subjects for semester " + semester);
+                return Promise.reject("Unable to fetch subjects for semester " + semester);
+            } else {
+                return Promise.resolve({semester: semester, classes: classesQuery.data.data.classes});
+            }
+        }));
+    }));
+
+    console.log("Net requests done!");
+
+    // a collection of classes documents to be added
+    // of the format of a map Class Title -> {classInfo, [Semester1, Semester2...]}
+    let classesInformation = new Map<string, any>();
+    
+    // update information about the class
+    let upsert = (classInfo, semester) => {
+        let info = classesInformation.get(classInfo.titleLong);
+        if (info) {
+            const newSemesters = info.semesters.includes(semester) ? info.semesters : info.semesters.concat([semester]);
+            classesInformation.set(classInfo.titleLong, {classInfo: info.classInfo, semesters: newSemesters});
+        } else {
+            classesInformation.set(classInfo.titleLong, {classInfo: classInfo, semesters: [semester]} );
+        }
+    }
+
+    semesterClasses.forEach(semester => semester.forEach(subjectData => subjectData.classes.forEach(cl => upsert(cl, subjectData.semester))));
+
+    // a list of new classes to be inserted into the db
+    let classDocs = [];
+    for (const [title, details] of classesInformation.entries()) {
+        // find any additional semester information present in the database and include it
+        let cl = await Classes.findOne({classTitle: title}).exec();
+        if (cl) {
+           cl.classSems.forEach(sem => !details.semester.includes(sem) ? details.semesters.push(sem) : 0);
+           await Classes.updateOne({classTitle: title}, {$set: {classSems: details.semesters}});
+        } else {
+            classDocs.push(new Classes({
+                classSub : details.classInfo.subject.toLowerCase(),
+                classNum : details.classInfo.catalogNbr,
+                classTitle : title,
+                classPrereq : [],
+                classFull: details.classInfo.subject.toLowerCase() + " " + details.classInfo.catalogNbr + " " + title,
+                classSems: details.semesters
+            }));
+        }
+    }
+
+    let classRes = await Classes.collection.insertMany(classDocs);
+
+    if (classRes.result.ok != 1) {
+        throw new Error("Unable to write classes to db!");
+    } else {
+        console.log("Done writing " + classRes.result.n + " classes to db!");
+    }
+
+    console.log(await Classes.find({}).exec());
```

Comment:
> Currently, it seems that the old approach is faster than the batched writing. This will be investigated further.

@khemritolya Is it possible that it's caused by this line in the new approach?

Reply:
It very well could be. I'll look into it when I get around to doing a proper analysis over the weekend.